README.md
# toGoogleStyle.js

### Click here for English (comming soon?)

## Yêu cầu
### jQuery (nên thêm kiểu này)
```html
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha384-vk5WoKIaW/vJyUAd9n/wmopsmNhiy+L2Z+SBxGYnUkunIxVxAv/UtMOhba/xskxh" crossorigin="anonymous"></script>
```

### Tải phông chữ này xuống
A. Bắt buộc<br>
• [GoogleSans-Bold](https://github.com/DELNEGEND/toGoogleStyle.js/blob/master/GoogleSans-Bold.otf "Ấn vào để download")

B. Không bắt buộc<br>
• [GoogleSans-Regular](https://github.com/DELNEGEND/toGoogleStyle.js/blob/master/GoogleSans-Regular.otf "Ấn vào để download")<br>
• [GoogleSans-Italic](https://github.com/DELNEGEND/toGoogleStyle.js/blob/master/GoogleSans-Italic.otf "Ấn vào để download")

## Thiết lập

### Thêm script toGoogleStyle này vào DOM (nên dùng cách này)
```html
<script src="https://cdn.jsdelivr.net/gh/delnegend/toGoogleStyle.js@1.0/toGoogleStyle.min.js" integrity="sha384-CsojGvi0BrX/Os+I75n6Tb3/ZgcazqptVr+FZjWCoxTPSXAC8ztx4/X8h4sbVvMW" crossorigin="anonymous"></script>
```

#### Hoặc
```html
<script src="https://dngnd.ml/toGoogleStyleJS"></script>
```

### Thêm font vào HTML

#### 1. Copy file phông chữ vào cùng thư mục với file HTML hoặc thư mục gốc trên server/hosting

#### 2. Add font vào DOM

- Cách 1: sử dụng [**loadExternalFont**](https://github.com/DELNEGEND/loadExternalFont.js "loadExternalFont repository"), ấn vào để xem chi tiết.
- Cách 2: sử dụng CSS

```css
@font-face {
  font-family: Google Sans;
  src: url('./GoogleSans-Bold'),
       url('./GoogleSans-Regular'),
       url('./GoogleSans-Italic');
}
```

## Hướng dẫn sử dụng

**1. Tự "bịa" ra 1 class nào đó rồi gán nó vào tất cả những thẻ div/span muốn biến hoá thành Google style, ví dụ như**

```html
<div class=".j2team_chang_han_the">Google</div>
<span class=".j2team_chang_han_the">Lorem Ipsum gì gì đấy...</span>
```

**2. Trong phần footer**

```html
<script>
toGoogleStyle('<tên class>','<cỡ chữ>','<kiểu chữ>')
<script>
```

- Tên class (dạng string, bắt buộc): .j2team_chang_han_the

- Cỡ chữ (dạng integer, không bắt buộc): mặc định là 40

- Kiểu chữ (dạng string, không bắt buộc): mặc định là 'bold', có thể để thành 'regular', 'italic' hoặc 'bolditalic' nhưng phải cài thêm 2 font đã đề cập trên

  *Ví dụ*
```html
<script>
toGoogleStyle('.abcxyz')
toGoogleStyle('.aaaaaaaa',69)
toGoogleStyle('.j2team_chang_han_the',96,'italic')
</script>
```

CHANGELOG.md
## Update 1.1 22.12.2019
• Loại bỏ loadExternalFont.js khỏi code

## Release 1.0 22.12.2019
• Phát hành bản đầu tiên

toGoogleStyle.js
// Main function
const _toGoogleStyle = selector => {

  // Take content inside div/span.selector and convert it into string, I called this the selected string
  let inputString = $(selector).html(),
    // Temporary array that will be converted into string to replace the selected string
    finalHTMLOutput = [],
    // Color list
    colorList = ["#4285F4", "#EA4335", "#FBBC05", "#4285F4", "#34A853", "#EA4335"];

  // Pointer of the colorList
  let colorIndex = 0;

  // Loop for each letter in the selected string
  for (let i = 0; i < inputString.length; i++) {
    // Incrementing the color pointer
    colorIndex++;

    // Reset the pointer to 0 when it hit the final value in the color list
    if (colorIndex == 6) (colorIndex -= 6);

    // Adding span and color to each individual letter
    let currOutLetter = `<span style="color: ${colorList[colorIndex]}">${inputString.charAt(i)}</span>`;
    
    // Push the individual letter to the temporary array
    finalHTMLOutput.push(currOutLetter);
  }
  // Finally join all values in that array into one single string and replace the selected string
  $(selector).html(finalHTMLOutput.join(""));
};

// Global style config like font, font-size, font-weight, font-style
const googleStyleConfig = (selector, fontSize, fontStyle) => {
  let weightOrDeco, _fw = 'font-weight',
    _td = 'text-decoration';
  switch (fontStyle) {
    case 'regular':
      weightOrDeco = '';
      break;
    case 'bold':
      weightOrDeco = `${_fw}:bold;`;
      break;
    case 'italic':
      weightOrDeco = `${_td}:italic;`;
      break;
    case 'bolditalic':
      weightOrDeco = `${_fw}:bold;${_td}:italic;`;
      break;
    default:
      weightOrDeco = "";
      break;
  }

  // Completed CSS string
  let googleStyleGlobalStylesheet = `${selector}>span{font-family:Google Sans;${weightOrDeco}font-size:${fontSize}px}`;

  // Duh
  let addCssToHTML = css => {
    let head = document.head || document.getElementsByTagName('head')[0],
      style = document.createElement('style');
    head.appendChild(style);
    style.type = 'text/css';
    if (style.styleSheet) {
      style.styleSheet.cssText = css;
    } else {
      style.appendChild(document.createTextNode(css));
    }
  };
  addCssToHTML(googleStyleGlobalStylesheet);

};

/*

If we include googleStyleConfig into _toGoogleStyle, there'll be 2 problems:
Nếu nhét hàm googleStyleConfig vào _toGoogleStyle và dùng luôn hàm đó pass class vào, sẽ xảy ra 2 vấn đề
1. Duplicated googleStyleConfig when the .google_style appear more than 1 time in DOM
2. All tags with .google_style class will have the same innerHTML

Example
<div.google_style> ABC </div>
<div.google_style> 123 </div>
<div.google_style> XYZ </div>
After running _toGoogleStyle('google_style')
<div.google_style> ABC </div>
<div.google_style> ABC </div>
<div.google_style> ABC </div>

The solution I know is to seperate the styleConfig(), passing class AND order into _toGoogleStyle one by one (can be done with .each() in jQuery)

*/
const toGoogleStyle = (selector, fontSize = 40, fontStyle = 'bold') => {
  $(selector).each((index, val) => _toGoogleStyle(val));
  googleStyleConfig(selector, fontSize, fontStyle);
};

// toGoogleStyle('.google_style');

toGoogleStyle_v1.0.js
// load Google Sans font from local
loadExternalFont(["Google Sans", './GoogleSans-Regular.otf', './GoogleSans-Bold.otf', './GoogleSans-Italic.otf']);

// Magics
const _toGoogleStyle = selector => {

  // Take content inside div/span.selector and convert it into string, I called this the selected string
  let inputString = $(selector).html(),
    // Temporary array that will be converted into string to replace the selected string
    finalHTMLOutput = [],
    // Color list
    colorList = ["#4285F4", "#EA4335", "#FBBC05", "#4285F4", "#34A853", "#EA4335"];

  // Pointer of the colorList
  let colorIndex = 0;

  // Loop for each letter in the selected string
  for (let i = 0; i < inputString.length; i++) {
    // Incrementing the color pointer
    colorIndex++;

    // Reset the pointer to 0 when it hit the final value in the color list
    if (colorIndex == 6) (colorIndex -= 6);

    // Adding span and color to each individual letter
    let currOutLetter = `<span style="color: ${colorList[colorIndex]}">${inputString.charAt(i)}</span>`;
    
    // Push the individual letter to the temporary array
    finalHTMLOutput.push(currOutLetter);
  }
  // Finally join all values in that array into one single string and replace the selected string
  $(selector).html(finalHTMLOutput.join(""));
};

// Global style config like font, font-size, font-weight, font-style
const googleStyleConfig = (selector, fontSize, fontStyle) => {
  let weightOrDeco, _fw = 'font-weight',
    _td = 'text-decoration';
  switch (fontStyle) {
    case 'regular':
      weightOrDeco = '';
      break;
    case 'bold':
      weightOrDeco = `${_fw}:bold;`;
      break;
    case 'italic':
      weightOrDeco = `${_td}:italic;`;
      break;
    case 'bolditalic':
      weightOrDeco = `${_fw}:bold;${_td}:italic;`;
      break;
    default:
      weightOrDeco = "";
      break;
  }

  // Completed CSS string
  let googleStyleGlobalStylesheet = `${selector}>span{font-family:Google Sans;${weightOrDeco}font-size:${fontSize}px}`;

  // Duh
  let addCssToHTML = css => {
    let head = document.head || document.getElementsByTagName('head')[0],
      style = document.createElement('style');
    head.appendChild(style);
    style.type = 'text/css';
    if (style.styleSheet) {
      style.styleSheet.cssText = css;
    } else {
      style.appendChild(document.createTextNode(css));
    }
  };
  addCssToHTML(googleStyleGlobalStylesheet);

};

/*

If we include googleStyleConfig into _toGoogleStyle, there'll be 2 problems:
Nếu nhét hàm googleStyleConfig vào _toGoogleStyle và dùng luôn hàm đó pass class vào, sẽ xảy ra 2 vấn đề
1. Duplicated googleStyleConfig when the .google_style appear more than 1 time in DOM
2. All tags with .google_style class will have the same innerHTML

Example
<div.google_style> ABC </div>
<div.google_style> 123 </div>
<div.google_style> XYZ </div>
After running _toGoogleStyle('google_style')
<div.google_style> ABC </div>
<div.google_style> ABC </div>
<div.google_style> ABC </div>

The solution I know is to seperate the styleConfig(), passing class AND order into _toGoogleStyle one by one (can be done with .each() in jQuery)

*/
const toGoogleStyle = (selector, fontSize = 40, fontStyle = 'bold') => {
  $(selector).each((index, val) => _toGoogleStyle(val));
  googleStyleConfig(selector, fontSize, fontStyle);
};

// toGoogleStyle('.google_style');